### PR TITLE
Add multiplex

### PR DESCRIPTION
**Pull request by eira-fransham**
Multiplex implementation. It's not completely finished, as such, but it's a functional implementation and shows that you don't need a "driver" function to write a protocol like this. After writing it this way I do wonder whether or not it would be better to use a "driver" function, since this requires unnecessary synchronisation overhead, although the fact that a lot of code is required to get anything done is not a function of the coordination. Another alternative would be to use `Cell`s for the single-threaded case and allow the user to choose whether or not to incur the synchronisation burden necessary to synchronise.

### TODO:

* Currently streams are closed locally when they are dropped (and therefore packets are dropped), but the remote endpoint is not notified. This isn't necessary in order to satisfy the protocol but is part of being a good network citizen.
* As mentioned above, we can do better optimisations for single-threaded async code.
* Currently IDs are generated by hashing an increasing nonce in order to avoid collisions between the two endpoints and by different instances of `OutboundStream`. This wastes a couple bytes over using small increasing nonces.

### CODE REVIEW

File: multiplex-rs/src/lib.rs
Comment (tomaka) [inline multiplex-rs/src/lib.rs]: Typo?

**Comment by folsen** [comment]
Do you want to fix those TODO's before review & merge or is this good to review now?

**Comment by tomaka** [comment]
Don't forget to add the license header to each file, and change the author in the Cargo.toml.

**Comment by Vurich** [comment]
Closes #8

File: multiplex-rs/src/read.rs
Comment (tomaka) [inline multiplex-rs/src/read.rs]: Same problem here

File: multiplex-rs/src/shared.rs
Comment (tomaka) [inline multiplex-rs/src/shared.rs]: Since you're using `unwrap_or_default()` when accessing these two fields anyway, what about initializing them to a default value instead of using `Option`?

**Review by tomaka — CHANGES_REQUESTED** [review]

File: multiplex-rs/src/shared.rs
Comment (Vurich) [reply multiplex-rs/src/shared.rs]: Because "was deliberately opened" + "now closed" is different to "never opened", but I can't remember the specifics as to why. Maybe it can be made a set now.

File: multiplex-rs/src/write.rs
Comment (Vurich) [reply multiplex-rs/src/write.rs]: Maybe I can return `Ok(0)`? I don't know what the right behaviour is here, `Ok(0)` is supposed to imply EOF

File: multiplex-rs/src/shared.rs
Comment (Vurich) [reply multiplex-rs/src/shared.rs]: Because I want to take ownership of them and don't want to construct a default value every call to `poll` (in order to use `mem::replace`).

File: multiplex-rs/src/shared.rs
Comment (Vurich) [reply multiplex-rs/src/shared.rs]: Almost certainly not, I'm just a performance pedant.

File: circular-buffer/src/lib.rs
Comment (rphmeier) [inline circular-buffer/src/lib.rs]: `OwnedSlice`'s destructor drops all elements in place, even those which have already been read out by the iterator

Comment (rphmeier) [reply circular-buffer/src/lib.rs]: (can I just say it again: please don't use any unsafe code at all!)

File: circular-buffer/src/lib.rs
Comment (rphmeier) [inline circular-buffer/src/lib.rs]: i am worried about the prospect of manually dropping uninitialized memory. it should rather drop all known-to-be-live elements in the destructor

Comment (Vurich) [reply circular-buffer/src/lib.rs]: I mean that the value of `B` must be manually dropped, as the default drop will drop all elements. What I do in the destructor is destroy only the known-to-be-live elements, just as you say.

File: futures-mutex/src/lib.rs
Comment (tomaka) [inline futures-mutex/src/lib.rs]: Since we're taking by `self`, can't we just return `self.inner`?

File: multiplex-rs/src/lib.rs
Comment (tomaka) [inline multiplex-rs/src/lib.rs]: Does that work now, or are there still conflicts with substream IDs?

File: multiplex-rs/src/lib.rs
Comment (tomaka) [inline multiplex-rs/src/lib.rs]: I don't think this function and `listen` (just below) are necessary?

File: multistream-select/src/dialer_select.rs
Comment (tomaka) [inline multistream-select/src/dialer_select.rs]: I know these `println!s` are mine, but it would be nice to remove them.

**Review by tomaka — CHANGES_REQUESTED** [review]

**Comment by tomaka** [comment]
@rphmeier If you agree, I would suggest merging this in the not-so-distant future (eg. today would be great). We have tons of stuff blocked by this pull request.

File: multistream-select/src/dialer_select.rs
Comment (Vurich) [reply multistream-select/src/dialer_select.rs]: I thought I'd got rid of this, it must have been as part of the commit I rolled back

File: multiplex-rs/src/lib.rs
Comment (Vurich) [reply multiplex-rs/src/lib.rs]: It works when we're the dialer, it _should_ work when we're the server too.

File: multiplex-rs/src/lib.rs
Comment (Vurich) [reply multiplex-rs/src/lib.rs]: I use them for tests and they're useful to manually construct a `Multiplex` if you're using it without `libp2p`

**Review by tomaka — APPROVED** [review]

**Comment by tomaka** [comment]
Merging in order to get things moving.

**Pull request merged**
